File: rock-paper-scissors/rock-paper-scissors.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <random>
#include <limits> // Required for numeric_limits

char getComputerChoice() {
    std::vector<char> choices = {'R', 'P', 'S'};
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> distrib(0, choices.size() - 1);
    return choices[distrib(gen)];
}

std::string determineWinner(char userChoice, char computerChoice, int& userScore, int& computerScore) {
    if (userChoice == computerChoice) {
        return "It's a tie!";
    } else if (
        (userChoice == 'R' && computerChoice == 'S') ||
        (userChoice == 'P' && computerChoice == 'R') ||
        (userChoice == 'S' && computerChoice == 'P')
    ) {
        userScore++;
        return "You win this round!";
    } else {
        computerScore++;
        return "Computer wins this round!";
    }
}

int main() {
    int userScore = 0;
    int computerScore = 0;
    int roundsToWin = 0;

    while (true) {
        std::cout << "Enter the number of rounds to win (an odd integer 3 or greater): ";
        std::cin >> roundsToWin;

        if (std::cin.fail() || roundsToWin < 3 || roundsToWin % 2 == 0) {
            std::cout << "Invalid input. Please enter an odd integer 3 or greater." << std::endl;
            std::cin.clear(); // Clear error flags
            // Ignore rest of the invalid input line
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        } else {
            // Ignore the rest of the line after valid integer input
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            break;
        }
    }

    std::cout << "First to win " << roundsToWin << " rounds wins the game!" << std::endl;

    while (userScore < roundsToWin && computerScore < roundsToWin) {
        std::string userInput;
        char userChoice;

        while (true) {
            std::cout << "Enter your choice (R for Rock, P for Paper, S for Scissors): ";
            std::getline(std::cin, userInput);

            if (userInput.length() == 1 && (userInput[0] == 'R' || userInput[0] == 'P' || userInput[0] == 'S')) {
                userChoice = userInput[0];
                break;
            } else {
                std::cout << "Invalid input. Please enter R, P, or S." << std::endl;
            }
        }

        char computerChoice = getComputerChoice();
        std::cout << "You chose: " << userChoice << std::endl;
        std::cout << "Computer chose: " << computerChoice << std::endl;

        std::string result = determineWinner(userChoice, computerChoice, userScore, computerScore);
        std::cout << result << std::endl;
        std::cout << "Score: You " << userScore << " - " << computerScore << " Computer" << std::endl;
    }

    std::cout << "\n--- Game Over ---" << std::endl;
    if (userScore > computerScore) {
        std::cout << "Congratulations! You won the game!" << std::endl;
    } else if (computerScore > userScore) {
        std::cout << "Computer won the game!" << std::endl;
    } else {
        std::cout << "The game is a tie!" << std::endl;
    }
    std::cout << "Final Score: You " << userScore << " - " << computerScore << " Computer" << std::endl;

    return 0;
}

File: rock-paper-scissors/rock-paper-scissors.py
import random

def get_computer_choice():
    return random.choice(['R', 'P', 'S'])

def determine_winner(user_choice, computer_choice):
    global user_score, computer_score
    if user_choice == computer_choice:
        return "It's a tie!"
    elif (user_choice == 'R' and computer_choice == 'S') or \
         (user_choice == 'P' and computer_choice == 'R') or \
         (user_choice == 'S' and computer_choice == 'P'):
        user_score += 1
        return "You win this round!"
    else:
        computer_score += 1
        return "Computer wins this round!"

user_score = 0
computer_score = 0
rounds_to_win = 0

while True:
    try:
        rounds_input = input("Enter the number of rounds to win (an odd integer 3 or greater): ")
        rounds_to_win = int(rounds_input)
        if rounds_to_win >= 3 and rounds_to_win % 2 != 0:
            break
        else:
            print("Invalid input. Please enter an odd integer 3 or greater.")
    except ValueError:
        print("Invalid input. Please enter a number.")

print(f"First to win {rounds_to_win} rounds wins the game!")

while user_score < rounds_to_win and computer_score < rounds_to_win:
    user_choice = input("Enter your choice (R for Rock, P for Paper, S for Scissors): ").upper()

    if user_choice not in ['R', 'P', 'S']:
        print("Invalid input. Please enter R, P, or S.")
        continue

    computer_choice = get_computer_choice()
    print(f"You chose: {user_choice}")
    print(f"Computer chose: {computer_choice}")

    result = determine_winner(user_choice, computer_choice)
    print(result)
    print(f"Score: You {user_score} - {computer_score} Computer")

print("\n--- Game Over ---")
if user_score > computer_score:
    print("Congratulations! You won the game!")
elif computer_score > user_score:
    print("Computer won the game!")
else:
    print("The game is a tie!")
print(f"Final Score: You {user_score} - {computer_score} Computer")
